### PR TITLE
Add conversions to Linkedin Ads Transform package

### PR DESCRIPTION
**Pull request by fivetran-avinash**
## PR Overview
**This PR will address the following Issue/Feature:** Adding conversion support

**This PR will result in the following new package version:**

Although no breaking changes are introduced, the PR is significantly robust enough to require a version bump.

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

## Feature Updates: Conversion Support!
We have added more robust support for conversions in our data models by doing the following: 
- Adds a `conversion_value_by_local_currency` field to each `_report` end model, representing the value of conversions that occurred on each day for each campaign/campaign_group/creative/url/account.
- Created a `linkedin_ads__conversion_fields` variable to pass through and transform additional conversion value metrics into their aggregate sums.
  - Set current variable defaults in the `dbt_project.yml` to bring in the most used conversion fields `external_website_conversions` and `one_click_leads` by default.
  - Instructions on how to set your own fields [are available in the README](https://github.com/fivetran/dbt_linkedin/blob/main/README.md#adding-in-conversion-fields-variable).
- Adds a `total_conversions` metric in our end models to track all conversions being brought in by the `linkedin_ads_conversion_fields` variable.

## Documentation Update 
- Documents how to set your own passthrough fields with the variable `linkedin_ads__conversion_fields` [in the README](https://github.com/fivetran/dbt_linkedin_source/blob/main/README.md#adding-in-conversion-fields-variable).
- [Created a DECISIONLOG](https://github.com/fivetran/dbt_linkedin/blob/main/DECISIONLOG.md#best-practices-with-configuring-linkedin-ads-conversion-fields-variable) to describe best practices in configuring the `linkedin_ads__conversion_fields` variable. 

## Under the Hood
- Added a new [version](https://github.com/fivetran/dbt_linkedin_ads/blob/main/macros/linkedin_ads_persist_pass_through_columns.sql) of the `persist_pass_through_columns()` [macro](https://github.com/fivetran/dbt_fivetran_utils/blob/v0.4.10/macros/persist_pass_through_columns.sql) in which we can include `coalesces` and properly check between conversion field values and the existing passthrough column.
- Updated the PR templates to align with our most up-to-date standards.
- Included auto-releaser GitHub Actions workflow to automate future releases.
- Addition of integrity and consistency validation tests within `integration_tests` for the Linkedin transformation models.
- Updated `linkedin_ad_analytics_by_creative_data` seed file with relevant conversion fields for more robust testing. 

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [ ] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [NA] The appropriate issue has been linked, tagged, and properly assigned.
- [x] All necessary documentation and version upgrades have been applied.
    
- [x] docs were regenerated (unless this PR does not include any code or yml updates).
- [x] BuildKite integration tests are passing.
- [x] Detailed validation steps have been provided below.

### Detailed Validation
Please share any and all of your validation steps:

Validation tests passing! [Mostly mirroring Jamie's validations in Facebook.](https://github.com/fivetran/dbt_facebook_ads/pull/43)

Screenshot 2024-06-25 at 11 25 25 AM

### If you had to summarize this PR in an emoji, which would it be?
 
🥼

### CODE REVIEW

**Review by fivetran-jamie — COMMENTED** [review]
looks GREAT! just doc comments really

File: README.md [hunk 1]
```diff
@@ -91,25 +91,44 @@ vars:
     linkedin_ads__use_local_currency: True # false by default -- uses USD
 ```
 
+Note: It should be noted that conversions are only reported in the local currency and are brought into our models as `conversion_value_in_local_currency`.
+
 ### Passing Through Additional Metrics
 By default, this package will select `clicks`, `impressions`, and `cost` from the source reporting tables to store into the staging models. If you would like to pass through additional metrics to the staging models, add the below configurations to your `dbt_project.yml` file. These variables allow for the pass-through fields to be aliased (`alias`) if desired, but not required. Use the below format for declaring the respective pass-through variables:
```
Comment (fivetran-jamie) [inline README.md:97]:
```suggestion
By default, this package will select `clicks`, `impressions`, `cost`, `conversion_value_in_local_currency`, and `total_conversions` (as well as fields set via `linkedin_ads__conversion_fields` in the next section) from the source reporting tables to store into the staging models. If you would like to pass through additional metrics to the staging models, add the below configurations to your `dbt_project.yml` file. These variables allow for the pass-through fields to be aliased (`alias`) if desired, but not required. Use the below format for declaring the respective pass-through variables:
```

File: DECISIONLOG.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+## Best Practices with Configuring LinkedIn Ads Conversion Fields Variable
+The `linkedin_ads__conversion_fields` variable is designed for end users to properly measure the conversions at the proper level of granularity. By default, we use `external_website_conversions` and `one_click_leads` as they are arguably the most used conversion measures, and fulfill entirely separate objectives as conversions (Website Conversion and Lead Generation respectively). 
+
+However, if you decide to configure your own conversion field variable fields, we highly recommend that you bring in conversions at the proper level of segmentation, so there aren't conversions that belong to multiple fields you bring in.
+
+### Bad Practice
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply DECISIONLOG.md:6]: Updated.

File: DECISIONLOG.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+```yml
+# dbt_project.yml
+vars:
+    linkedin_ads__conversion_fields: ['external_website_conversions', 'external_website_pre_click_conversions', 'external_website_post_click_conversions']
+```
+
+`external_website_conversions` is comprised of both `external_website_pre_click_conversions` and `external_website_post_click_conversions`, so `total_conversions` in the end models would be double counting conversions. 
+
+### Good Practice
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply DECISIONLOG.md:16]: Updated

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,29 @@
...
+[PR #36](https://github.com/fivetran/dbt_linkedin/pull/36) includes the following updates:
+
+## Feature Updates: Conversion Support!
+We have added more robust support for conversions in our data models by doing the following: 
+- Adds a `conversion_value_by_local_currency` field to each `_report` end model, representing the value of conversions that occurred on each day for each campaign/campaign_group/creative/url/account.
+- Created a `linkedin_ads__conversion_fields` variable to pass through and transform additional conversion value metrics into their aggregate sums.
+  - Set current variable defaults in the `dbt_project.yml` to bring in the most used conversion fields `external_website_conversions` and `one_click_leads` by default.
+  - Instructions on how to set your own fields [are available in the README](https://github.com/fivetran/dbt_linkedin/blob/main/README.md#adding-in-conversion-fields-variable).
+- Adds a `total_conversions` metric in our end models to track all conversions being brought in by the `linkedin_ads_conversion_fields` variable.
+
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply CHANGELOG.md:11]: Updated.

File: README.md [hunk 5]
```diff
@@ -91,25 +91,44 @@ vars:
     linkedin_ads__use_local_currency: True # false by default -- uses USD
 ```
 
+Note: It should be noted that conversions are only reported in the local currency and are brought into our models as `conversion_value_in_local_currency`.
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply README.md:94]: Updated.

File: README.md [hunk 6]
```diff
@@ -91,25 +91,44 @@ vars:
...
+        - name: "another_one"
+          transform_sql: "coalesce(another_one, 0)" # reference the `name` here if you're not using an alias
         - name: "that_field"
     linkedin_ads__creative_passthrough_metrics: # pulls from ad_analytics_by_creative
         - name: "new_custom_field"
           alias: "custom_field"
         - name: "unique_int_field"
 ```
+
+>**Note** Please ensure you exercised due diligence when adding metrics to these models. The metrics added by default (clicks, impressions, and cost) have been vetted by the Fivetran team maintaining this package for accuracy. There are metrics included within the source reports, for example metric averages, which may be inaccurately represented at the grain for reports created in this package. You will want to ensure whichever metrics you pass through are indeed appropriate to aggregate at the respective reporting levels provided in this package.
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply README.md:117]: Updated

File: README.md [hunk 7]
```diff
@@ -91,25 +91,44 @@ vars:
...
+
+### Adding in Conversion Fields Variable
+We introduced support for conversion fields in our `report` data models in the [v0.9.0 release](https://github.com/fivetran/dbt_linkedin/releases/tag/v0.9.0) of the package, but customers might have been bringing in these conversion fields earlier using the passthrough fields variables. To avoid duplicate column errors, we have introduced the `linkedin_ads__conversion_fields` variable.
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply README.md:120]: Added

File: README.md [hunk 8]
```diff
@@ -91,25 +91,44 @@ vars:
...
+
+### Adding in Conversion Fields Variable
+We introduced support for conversion fields in our `report` data models in the [v0.9.0 release](https://github.com/fivetran/dbt_linkedin/releases/tag/v0.9.0) of the package, but customers might have been bringing in these conversion fields earlier using the passthrough fields variables. To avoid duplicate column errors, we have introduced the `linkedin_ads__conversion_fields` variable. 
+
+By default, these data models are set to bring in `external_website_conversions` and `one_click_leads`, the most used conversion fields. If you would like to modify which conversion fields you are bringing in, you can set them in the `dbt_project.yml`.
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply README.md:122]: Updated.

File: README.md [hunk 9]
```diff
@@ -91,25 +91,44 @@ vars:
...
+
+```yml
+# dbt_project.yml
+vars:
+    linkedin_ads__conversion_fields: ['external_website_pre_click_conversions', 'one_click_leads', 'external_website_post_click_conversions', 'landing_page_clicks']
+```
+
+Make sure to follow best practices in configuring fields in the conversion field variables! [See the DECISIONLOG for more details](https://github.com/fivetran/dbt_linkedin/blob/main/DECISIONLOG.md#best-practices-with-configuring-linkedin-ads-conversion-fields-variable). 
+
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply README.md:131]: Updated

File: integration_tests/package-lock.yml [hunk 10]
```diff
@@ -0,0 +1,11 @@
+packages:
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply integration_tests/package-lock.yml:1]: Strange, i thought it was already added. Updated.

File: models/linkedin_ads__campaign_report.sql [hunk 11]
```diff
@@ -60,12 +65,16 @@ final as (
         campaign.run_schedule_end_at,
         campaign.last_modified_at,
         campaign.created_at,
+        report.total_conversions,
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/linkedin_ads__campaign_report.sql:68]: Oh wow how did I miss this 🤦🏽 . Thanks and fixed.

File: models/linkedin_ads__creative_report.sql [hunk 12]
```diff
@@ -53,9 +58,13 @@ final as (
         account.currency,
         creative.last_modified_at,
         creative.created_at,
+        report.total_conversions,
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/linkedin_ads__creative_report.sql:61]: Fixed.

File: models/linkedin_ads__url_report.sql [hunk 13]
```diff
@@ -56,12 +61,16 @@ final as (
         account.account_id,
         account.account_name,
         account.currency,
+        report.total_conversions,
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/linkedin_ads__url_report.sql:64]: Fixed.

File: models/linkedin_ads__campaign_group_report.sql [hunk 14]
```diff
@@ -43,12 +48,16 @@ final as (
         campaign_group.run_schedule_end_at,
         campaign_group.last_modified_at,
         campaign_group.created_at,
+        report.total_conversions,
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/linkedin_ads__campaign_group_report.sql:51]: Fixed.

[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/linkedin_ads__campaign_group_report.sql:51]: I think `linkedin_ads__account_report` wasn't included but looking at the development results, this should be summed as well.

File: README.md (hunk 1 above)
Comment (fivetran-avinash) [reply README.md:97]: Updated

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,3 +1,29 @@
...
+## Under the Hood
+- Added a new [version](https://github.com/fivetran/dbt_linkedin_ads/blob/main/macros/linkedin_ads_persist_pass_through_columns.sql) of the `persist_pass_through_columns()` [macro](https://github.com/fivetran/dbt_fivetran_utils/blob/v0.4.10/macros/persist_pass_through_columns.sql) in which we can include `coalesces` and properly check between conversion field values and the existing passthrough column.
+- Updated the PR templates to align with our most up-to-date standards.
+- Included auto-releaser GitHub Actions workflow to automate future releases.
+- Addition of integrity and consistency validation tests within `integration_tests` for the Linkedin transformation models.
+- Updated `linkedin_ad_analytics_by_creative_data` seed file with relevant conversion fields for more robust testing. 
+
+## Contributors
+[Seer Interactive](https://www.seerinteractive.com/)
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply CHANGELOG.md:25]: Updated

File: models/linkedin.yml [hunk 16]
```diff
@@ -61,6 +61,12 @@ models:
           BUSINESS - This is the only value allowed when creating accounts through the API.
           ENTERPRISE - This value cannot be used to create accounts through the API and is reserved for accounts created by LinkedIn's internal ad operations systems.
 
+      - name: conversion_value_in_local_currency
+        description: The value generated by your conversions, displayed in your LinkedIn account's local currency, on the given day.
+
+      - name: total_conversions
+        description: The total conversions being brought in the conversion fields you select in your `linkedin_ads__conversion_fields`.
+
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/linkedin.yml:69]: Added.

File: models/linkedin.yml [hunk 17]
```diff
@@ -215,7 +221,13 @@ models:
...
       - name: currency
         description: The ISO 4217 currency code of the account, such as 'USD' for United States Dollar (default).
-  
+
+      - name: conversion_value_in_local_currency
+        description: The value generated by your conversions, displayed in your LinkedIn account's local currency, on the given day.
+
+      - name: total_conversions
+        description: The total conversions being brought in the conversion fields you select in your `linkedin_ads__conversion_fields`.
+
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/linkedin.yml:230]: Added

File: models/linkedin.yml [hunk 18]
```diff
@@ -286,6 +298,12 @@ models:
       - name: currency
         description: The ISO 4217 currency code of the account, such as 'USD' for United States Dollar (default).
 
+      - name: conversion_value_in_local_currency
+        description: The value generated by your conversions, displayed in your LinkedIn account's local currency, on the given day.
+
+      - name: total_conversions
+        description: The total conversions being brought in the conversion fields you select in your `linkedin_ads__conversion_fields`.
+
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/linkedin.yml:306]: Added

File: models/linkedin.yml [hunk 19]
```diff
@@ -382,6 +400,11 @@ models:
       - name: currency
         description: The ISO 4217 currency code of the account, such as 'USD' for United States Dollar (default).
 
+      - name: conversion_value_in_local_currency
+        description: The value generated by your conversions, displayed in your LinkedIn account's local currency, on the given day.
+
+      - name: total_conversions
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply models/linkedin.yml:406]: Added

**Comment by fivetran-avinash** [comment]
> looks GREAT! just doc comments really

@fivetran-jamie Ready for re-review!

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-jamie Ready for re-review!

**Review by fivetran-jamie — APPROVED** [review]
🥳 Looks great to me! I'll take on merging this in this/next sprint

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-jamie and @fivetran-avinash great work on these PRs! I only have the one callout to update the README section for the new conversion variable, but all other stress tests I performed look good! Once that final update is applied this should be good to go.

**Pull request merged**